feat: add `blas/ext/base/dcopy-within`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: missing_dependencies
  - task: lint_c_examples status: missing_dependencies
  - task: lint_c_benchmarks status: missing_dependencies
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/97

## Description

> What is the purpose of this pull request?

This pull request:

-   add `blas/ext/base/dcopy-within`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/97

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add blas/ext/base/dcopy-within

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
/stdlib help

headlessNode commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/README.md:
````diff
@@ -0,0 +1,302 @@
+var dcopyWithin = require( '@stdlib/blas/ext/base/dcopy-within' );
+```
+
+#### dcopyWithin( N, target, x, strideX, start, end )
````
Splitting the `target`, `start`, and `end` parameters is a bit awkward. Instead, how about

```
dcopyWithin( N, target, start, end, x, strideX )
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/README.md:
```diff
@@ -0,0 +1,302 @@
+
+var x = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0 ] );
+
+dcopyWithin( 3, 2, x, 1, 1, 2 );
```
A single value copy example isn't super informative. Set `N = x.length` and copy additional elements.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/README.md:
````diff
@@ -0,0 +1,302 @@
+```javascript
+var Float64Array = require( '@stdlib/array/float64' );
+
+var x = new Float64Array( [ 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0 ] );
````
I'd make this a tad more obvious by zeroing the non-indexed values.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/README.md:
```diff
@@ -0,0 +1,302 @@
+var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
+
+// Copy within the view...
+dcopyWithin( 2, 0, x1, 1, 2, 5 );
```
A 2-element array is not super-informative.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/lib/ndarray.js:
```diff
@@ -0,0 +1,95 @@
+	// Resolve copy length:
+	cl = min( abs( end - start ), N );
+
+	// Create a temporary copy of source elements:
```
1. While always copying to a temporary array is "safe", it is not the most performant. Ideally, we'd determine whether we actually need to copy. E.g., see https://github.com/stdlib-js/stdlib/blob/447a1ea0a03fa642d8193bdefd6bcb2533f99148/lib/node_modules/%40stdlib/array/complex128/lib/main.js#L2007.

We only need to copy when writing into the target region would overwrite the values we are wanting to copy. It should be possible to determine this.

2. It's not clear why you are needing to use manual `for` loops once you actually perform the copy. This is exactly the use case of `dcopy`.ndarray. And if you use `dcopy`, you can get perf benefits due to hardware acceleration. In short, ensure you are appropriately leveraging what we already have rather than reinventing the wheel.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
docs: apply suggestions from code review

headlessNode committed:
refactor: negative test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/src/main.c:
```diff
@@ -0,0 +1,113 @@
+	FLG = 0;
+
+	// Resolve copy length:
+	if ( stdlib_base_abs( end - start ) < N ) {
```
We don't want to use `stdlib_base_abs` here as the input arg is converted to a double and these values may be `int64`.

headlessNode committed:
Merge branch 'stdlib-js:develop' into dcopy-within

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte ~~we're getting an error in C, apparently due to this: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/number/float16/ctor/include/stdlib/number/float16/ctor.h#L50~~

Edit: My bad. It wasn't that causing the error.

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: failing tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: failing tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: lint errors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
Merge branch 'develop' into dcopy-within

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: passed
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte Ready 🤞🏻

kgryte commented:
As discussed, need to add workspace array support.

headlessNode committed:
Merge branch 'develop' into dcopy-within

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: passed
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte Ready for review.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+}
+
+/**
+* Run a benchmark which tests `stdlib_strided_dcopy_within`.
```
```suggestion
* Run a benchmark.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+}
+
+/**
+* Run a benchmark which tests `stdlib_strided_dcopy_within_ndarray`.
```
```suggestion
* Run a benchmark.
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+	print_version();
+	count = 0;
+
+	// Run benchmarks:
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+		iter = ITERATIONS / pow( 10, i-1 );
+		if ( iter < 1 ) {
+			iter = 1;
+		}
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+}
+
+/**
+* Generates a random double on the interval [-100,100).
```
```suggestion
* Generates a random number on the interval [0,1).
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+/**
+* Generates a random double on the interval [-100,100).
+*
+* @return random double
```
```suggestion
* @return random number
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+*/
+static double rand_double( void ) {
+	int r = rand();
+	return -100.0 + (r/(double)RAND_MAX)*200.0;
```
```suggestion
	return (double)r / ( (double)RAND_MAX + 1.0 );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+*
+* @param iterations   number of iterations
+* @param len          array length
+* @return elapsed time in seconds
```
```suggestion
* @return             elapsed time in seconds
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+*
+* @param iterations   number of iterations
+* @param len          array length
+* @return elapsed time in seconds
```
```suggestion
* @return             elapsed time in seconds
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+}
+
+/**
+* Run a benchmark.
```
```suggestion
* Runs a benchmark.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+}
+
+/**
+* Run a benchmark.
```
```suggestion
* Runs a benchmark.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+			elapsed = benchmark1( iter, len );
+			print_results( "dcopy_within:len", elapsed );
+			printf( "ok %d benchmark finished\n", count );
+
```
```suggestion
		}
		for ( j = 0; j < REPEATS; j++ ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,207 @@
+	}
+	print_summary( count, count );
+
+	return 0;
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,201 @@
+* Prints benchmark results.
+*
+* @param prefix  comment prefix
+* @param elapsed elapsed time in seconds
```
```suggestion
* @param iterations   number of iterations
* @param elapsed      elapsed time in seconds
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,201 @@
+	printf( "  elapsed: %0.9f\n", elapsed );
+	printf( "  rate: %0.9f\n", rate );
+	printf( "  ...\n" );
+}
```
```suggestion
static void print_results( int iterations, double elapsed ) {
	double rate = (double)iterations / elapsed;
	printf( "  ---\n" );
	printf( "  iterations: %d\n", iterations );
	printf( "  elapsed: %0.9f\n", elapsed );
	printf( "  rate: %0.9f\n", rate );
	printf( "  ...\n" );
}
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,201 @@
+			count += 1;
+			printf( "# c::%s:len=%d\n", NAME, len );
+			elapsed = benchmark1( iter, len );
+			print_results( "dcopy_within:len", elapsed );
```
```suggestion
			print_results( iter, elapsed );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,201 @@
+			count += 1;
+			printf( "# c::%s:ndarray:len=%d\n", NAME, len );
+			elapsed = benchmark2( iter, len );
+			print_results( "dcopy_within_ndarray:len", elapsed );
```
```suggestion
			print_results( iter, elapsed );
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/benchmark.js:
```diff
@@ -0,0 +1,108 @@
+var pow = require( '@stdlib/math/base/special/pow' );
+var format = require( '@stdlib/string/format' );
+var dcopyWithin = require( './../lib/dcopy_within.js' );
+var pkg = require( './../package.json' ).name;
```
```suggestion
var pkg = require( './../package.json' ).name;
var dcopyWithin = require( './../lib/dcopy_within.js' );
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,108 @@
+var pow = require( '@stdlib/math/base/special/pow' );
+var format = require( '@stdlib/string/format' );
+var dcopyWithin = require( './../lib/ndarray.js' );
+var pkg = require( './../package.json' ).name;
```
```suggestion
var pkg = require( './../package.json' ).name;
var dcopyWithin = require( './../lib/ndarray.js' );
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/docs/types/index.d.ts:
```diff
@@ -0,0 +1,124 @@
+*/
+interface Routine {
+	/**
+	* Performs an in-place copy of elements within a double-precision floating-point array.
```
```suggestion
	* Performs an in-place copy of elements within a double-precision floating-point strided array.
```

Here and for every other description.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/examples/c/example.c:
```diff
@@ -0,0 +1,43 @@
+	// Create a workspace array:
+	double w[ 8 ];
+
+	// Specify the number of elements to copy:
```
```suggestion
	// Specify the number of indexed elements:
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/examples/c/example.c:
```diff
@@ -0,0 +1,43 @@
+	const int strideX = 1;
+	const int strideW = 1;
+
+	// Copy elements starting from index 1 to index 4:
```
```suggestion
	// Copy elements:
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/test/test.dcopy_within.native.js:
```diff
@@ -0,0 +1,304 @@
+var dcopyWithin = tryRequire( resolve( __dirname, './../lib/dcopy_within.native.js' ) );
+var opts = {
+	'skip': ( isBrowser || isError( dcopyWithin ) )
+};
```
```suggestion
var Float64Array = require( '@stdlib/array/float64' );
var tryRequire = require( '@stdlib/utils/try-require' );


// VARIABLES //

var dcopyWithin = tryRequire( resolve( __dirname, './../lib/dcopy_within.native.js' ) );
var opts = {
	'skip': ( dcopyWithin instanceof Error )
};
```

I am not sure what AI model you are using, but it is not good.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/test/test.js:
```diff
@@ -0,0 +1,82 @@
+	}
+});
+
+tape( 'if a native implementation is not available, the main export is a JavaScript implementation', function test( t ) {
```
```suggestion
tape( 'if a native implementation is not available, the main export is a JavaScript implementation', opts, function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/test/test.ndarray.native.js:
```diff
@@ -0,0 +1,306 @@
+var dcopyWithin = tryRequire( resolve( __dirname, './../lib/ndarray.native.js' ) );
+var opts = {
+	'skip': ( isBrowser || isError( dcopyWithin ) )
+};
```
```suggestion
var Float64Array = require( '@stdlib/array/float64' );
var isError = require( '@stdlib/assert/is-error' );


// VARIABLES //

var dcopyWithin = tryRequire( resolve( __dirname, './../lib/ndarray.native.js' ) );
var opts = {
	'skip': ( dcopyWithin instanceof Error )
};
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/test/test.ndarray.native.js:
```diff
@@ -0,0 +1,304 @@
+var resolve = require( 'path' ).resolve;
+var tape = require( 'tape' );
+var Float64Array = require( '@stdlib/array/float64' );
+var isError = require( '@stdlib/assert/is-error' );
```
```suggestion
var tryRequire = require( '@stdlib/utils/try-require' );
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/README.md:
```diff
@@ -0,0 +1,323 @@
+    // Create a workspace array:
+    double w[ 8 ];
+
+    // Specify the number of elements to copy:
```
```suggestion
    // Specify the number of indexed elements:
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/README.md:
```diff
@@ -0,0 +1,323 @@
+    const int strideX = 1;
+    const int strideW = 1;
+
+    // Copy elements starting from index 1 to index 4:
```
```suggestion
    // Copy elements:
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
Okay. I think the only thing left is updating all the descriptions. This PR needed much closer review before requesting official review. Many things should have been caught earlier on in the process as divergences from recently added packages.

headlessNode committed:
fix: docs descriptions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: passed
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
Merge branch 'develop' into dcopy-within

headlessNode committed:
fix: cleanup

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: na
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte Changes applied.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/lib/dcopy_within.native.js:
```diff
@@ -0,0 +1,59 @@
+/**
+* Performs an in-place copy of elements within a double-precision floating-point strided array.
+*
+* @private
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/lib/ndarray.native.js:
```diff
@@ -0,0 +1,61 @@
+/**
+* Performs an in-place copy of elements within a double-precision floating-point strided array.
+*
+* @private
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/dcopy-within/src/main.c:
```diff
@@ -0,0 +1,84 @@
+	if ( N <= 0 || strideX == 0 || target >= N ) {
+		return;
+	}
+	// Resolve the number of elements to copy (note: intentionally avoid `stdlib_base_min`, which operates on double-precision floating-point numbers)...
```
```suggestion
	// Resolve the number of elements to copy...
```

This aside is leftover LLM junk.

kgryte committed:
Apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
